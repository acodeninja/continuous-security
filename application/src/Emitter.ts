import EventEmitter2 from 'eventemitter2';

export class Emitter implements EventEmitter {
  private eventEmitter: EventEmitter2;

  constructor() {
    this.eventEmitter = new EventEmitter2({wildcard: true});
  }

  emit(event: string, ...values: Array<unknown>): void {
    this.eventEmitter.emit(event, ...values);
  }

  on(event: string, listener: (...values: Array<unknown>) => void): void {
    this.eventEmitter.on(event, listener);
  }
}
