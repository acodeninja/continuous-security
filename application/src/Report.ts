import {template, TemplateExecutor} from 'lodash';
import {basename, resolve} from 'path';
import {readFile, writeFile} from 'fs/promises';
import {Converter as Showdown} from 'showdown';
import {promisify} from 'util';
import {exec} from 'child_process';

import {Emitter} from './Emitter';
import {translate} from './DataSources/Translations';
import {References} from './DataSources/References';
import {makeTemporaryFolder} from './Helpers/Files';

import MarkdownTemplate from './assets/report.markdown.template.md';
import PDFTemplate from './assets/report.pdf.template.md';
import HTMLTemplate from './assets/report.html.template.md';
import HTMLTemplateWrapper from './assets/report.html.wrapper.html';

export class Report {
  private readonly templates: Record<string, TemplateExecutor>;
  private reports: Array<ScanReport> = [];
  private cached: ReportOutput;
  private emitter: Emitter;
  private references: References;
  private severityOrder = ['critical', 'high', 'moderate', 'low', 'info', 'unknown'];
  private reportFunctions: Record<string, (...args: Array<unknown>) => unknown> = {
    groupBy: (list: Array<unknown>, grouping: string) => {
      const groupedIssues =
        [{}, ...list].reduce((group: Record<string, Array<unknown>> = {}, item) => {
          if (item) {
            group[item[grouping]] = group[item[grouping]] ?? [];
            group[item[grouping]].push(item);
          }
          return group;
        });

      const asEntries = Object.entries(groupedIssues);

      asEntries.sort((a, b) => this.severityOrder.indexOf(a[0]) - this.severityOrder.indexOf(b[0]));

      return Object.fromEntries(asEntries);
    },
    capitalise: (words: string) =>
      words.split(' ')
        .map(w => w.charAt(0).toUpperCase() + w.slice(1))
        .join(' '),
  };

  constructor(emitter: Emitter) {
    this.emitter = emitter;
    this.references = new References(emitter);
    this.templates = {
      markdown: template(MarkdownTemplate),
      html: template(HTMLTemplate),
      pdf: template(PDFTemplate),
    };
  }

  addScanReport(report: ScanReport): void {
    this.reports.push(report);
  }

  async toObject(): Promise<ReportOutput> {
    if (this.cached) return this.cached;

    const issues = await this.getIssues();

    const overviewOfIssues = issues.map(i => i.references).flat()
      .filter(r => !!r.dataSourceSpecific.cwe)
      .filter((r, i, all) => all.map(a => a.label).indexOf(r.label) === i);

    const summaryImpacts: Record<string, Array<string>> = {};

    overviewOfIssues
      .map(({dataSourceSpecific}) =>
        dataSourceSpecific?.cwe?.consequences?.map(({scopeImpacts}) =>
          scopeImpacts,
        ),
      )
      .flat(2)
      .forEach(si => {
        summaryImpacts[si.scope] = summaryImpacts[si.scope] ?? [];
        if (si.impact && !summaryImpacts[si.scope].includes(si.impact)) {
          summaryImpacts[si.scope].push(si.impact);
        }
      });

    overviewOfIssues.sort((a, b) =>
      parseInt(a.label.replace('CWE-', '')) - parseInt(b.label.replace('CWE-', '')));

    const severities = issues.map(i => i.severity);

    this.cached = JSON.parse(JSON.stringify({
      title: `Security Report for ${basename(process.cwd())}`,
      date: new Date(),
      summaryImpacts: Object.entries(summaryImpacts).map(([scope, impacts]) => ({scope, impacts})),
      overviewOfIssues,
      issues,
      counts: {
        critical: severities.filter(s => s === 'critical').length,
        high: severities.filter(s => s === 'high').length,
        moderate: severities.filter(s => s === 'moderate').length,
        low: severities.filter(s => s === 'low').length,
        info: severities.filter(s => s === 'info').length,
        unknown: severities.filter(s => s === 'unknown').length,
        total: severities.length,
      },
    }), (key, value) => {
      if (
        String(value).length === 24 &&
        String(value).match(/\d{4}-\d{2}-\d{2}T(\d{2}:){2}\d{2}.\d{3}Z/)
      ) return new Date(value);

      if (key === 'code' || key === 'target' || key === 'name') return value;

      return typeof value === 'string' ? translate(value) : value;
    });

    return this.cached;
  }

  async toJSON(): Promise<[string, Buffer]> {
    const report = await this.toObject();
    this.emitter.emit('report:finished', '');

    return ['json', Buffer.from(JSON.stringify(report, null, 2))];
  }

  async toMarkdown(): Promise<[string, Buffer]> {
    const report = await this.toObject();
    this.emitter.emit('report:finished', '');

    return [
      'md',
      Buffer.from(this.templates.markdown({...report, functions: this.reportFunctions})),
    ];
  }

  async toHTML(): Promise<[string, Buffer]> {
    const report = await this.toObject();
    this.emitter.emit('report:finished', '');

    const convert = new Showdown();
    const converted = convert.makeHtml(this.templates.html({
      ...report,
      functions: this.reportFunctions,
    }));

    const html = HTMLTemplateWrapper.toString().replace('%%REPORT%%', converted);

    return ['html', Buffer.from(html)];
  }

  async toPDF(): Promise<[string, Buffer]> {
    const htmlCacheLocation = await makeTemporaryFolder('html');

    const report = await this.toObject();
    this.emitter.emit('report:finished', '');

    const convert = new Showdown();
    const converted = convert.makeHtml(this.templates.pdf({
      ...report,
      functions: this.reportFunctions,
    }));

    const html = HTMLTemplateWrapper.toString().replace('%%REPORT%%', converted);

    await writeFile(resolve(htmlCacheLocation, 'report.html'), html);

    const pdfPath = resolve(htmlCacheLocation, 'report.pdf');
    const htmlPath = resolve(htmlCacheLocation, 'report.html');

    // chrome --headless --disable-gpu --print-to-pdf https://www.chromestatus.com/

    try {
      await promisify(exec)(
        `google-chrome \
        --headless \
        --disable-gpu \
        --no-margins \
        --no-pdf-header-footer \
        --print-to-pdf="${pdfPath}" \
        ${htmlPath}`,
      );
    } catch (e) {
      console.log(e);
    }

    return ['pdf', Buffer.from(await readFile(pdfPath))];
  }

  async getReport(type: 'markdown' | 'json' | 'html' | 'pdf'): Promise<[string, Buffer]> {
    this.emitter.emit('report:started', `generating output report in ${type}`);
    switch (type) {
    case 'markdown':
      return await this.toMarkdown();
    case 'html':
      return await this.toHTML();
    case 'pdf':
      return await this.toPDF();
    case 'json':
      return await this.toJSON();
    }
  }

  private async getIssues(): Promise<Array<ReportOutputIssue>> {
    const withFoundBy = this.reports.map(r => r.issues.map(issue => ({
      ...issue,
      foundBy: r.scanner,
    }))).flat();

    const withExpandedReferences = [];

    for (const issue of withFoundBy) {
      withExpandedReferences.push({
        ...issue,
        references: issue.references ? await this.references.getAll(issue.references) : [],
      });
    }

    const withHighestSeverities = withExpandedReferences.map(i => {
      const severities = i.references
        .map((r: ReportOutputIssueReference) => r.dataSourceSpecific?.osv?.severity)
        .concat(i.severity)
        .filter((s: ReportOutputIssue['severity'] | undefined) => !!s);

      severities.sort((a, b) => this.severityOrder.indexOf(a) - this.severityOrder.indexOf(b));

      return {...i, severity: severities[0]};
    });

    return await Promise.all(withHighestSeverities.map(async issue => {
      return {
        ...issue,
        extracts: issue.extracts ? await Promise.all(issue.extracts.map(async extract => {
          const code = (await readFile(extract.path)).toString().split('\n');
          const start = Math.max(parseInt(extract.lines[0]) - 3, 0);
          const end = Math.min(parseInt(extract.lines[0]) + 2, code.length);
          const lines = new Array(code.slice(start, end).length)
            .fill(null)
            .map((_, i) => start + i + 1);

          return {
            ...extract,
            lines: extract.lines.map(line => `${line}`),
            code: code.slice(start, end).map((l, i) => {
              const length = lines[lines.length - 1].toString().length;
              return `${lines[i].toString().padStart(length)}| ${l}`;
            }).join('\n'),
          };
        })) : undefined,
      };
    }));
  }
}
