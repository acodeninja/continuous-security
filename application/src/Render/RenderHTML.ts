import {template} from 'lodash';

import {Report} from '../Report';
import {Emitter} from '../Emitter';
import {groupToObjectBy} from '../Helpers/Arrays';
import {capitalise} from '../Helpers/Strings';

import HTMLTemplate from '../assets/report.html.template.md';
import HTMLTemplateWrapper from '../assets/report.html.wrapper.html';
import {Converter as Showdown} from 'showdown';
import {timezone, toDate, toTime} from '../Helpers/Dates';

export class RenderHTML {
  private report: Report;
  private emitter: Emitter;

  constructor(report: Report, emitter: Emitter) {
    this.report = report;
    this.emitter = emitter;
  }

  async render(): Promise<Buffer> {
    this.emitter.emit('report:render:html:started');

    const report = await this.report.toObject();
    const convert = new Showdown();

    const innerReport = convert.makeHtml(template(HTMLTemplate)({...report, functions: {
      groupToObjectBy,
      capitalise,
      toDate,
      toTime,
      timezone,
    }}));

    const output = HTMLTemplateWrapper.toString().replace('%%REPORT%%', innerReport);

    this.emitter.emit('report:render:html:finished');

    return Buffer.from(output);
  }
}
