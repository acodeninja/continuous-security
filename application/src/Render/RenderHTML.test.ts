import {Report} from '../Report';
import {Emitter} from '../Emitter';
import {RenderHTML} from './RenderHTML';

import {emptyReport, fullReport} from '../../tests/fixtures/Reports';

jest.mock('../Helpers/Dates', () => ({
  toDate: jest.fn().mockReturnValue('01/04/2020'),
  toTime: jest.fn().mockReturnValue('01:30:10'),
  timezone: jest.fn().mockReturnValue('UTC+0'),
}));

describe('RenderHTML', () => {
  describe.each([
    ['empty', emptyReport],
    ['full', fullReport],
  ])('an %s report', (name, reportTemplate) => {
    const onReportRenderStarted = jest.fn();
    const onReportRenderFinished = jest.fn();
    const emitter = new Emitter();
    const report = new Report(emitter);
    const render = new RenderHTML(report, emitter);
    const reportToObject = jest.spyOn(report, 'toObject');
    let output: Buffer;

    beforeAll(async () => {
      emitter.on('report:render:html:started', onReportRenderStarted);
      emitter.on('report:render:html:finished', onReportRenderFinished);
      reportToObject.mockResolvedValue(reportTemplate);
      output = await render.render();
    });

    test('emits event report:render:html:started', () => {
      expect(onReportRenderStarted).toHaveBeenCalled();
    });

    test('emits event report:render:html:finished', () => {
      expect(onReportRenderFinished).toHaveBeenCalled();
    });

    test('calls Report.toObject', () => {
      expect(reportToObject).toHaveBeenCalledWith();
    });

    test('returns the expected report', () => {
      expect(output.toString()).toMatchSnapshot();
    });
  });
});
