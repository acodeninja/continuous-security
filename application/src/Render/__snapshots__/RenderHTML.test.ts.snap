// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderHTML an empty report returns the expected report 1`] = `
"<!doctype html>
<html lang="en-gb">
<head>
    <title></title>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@5.2.3/dist/css/bootstrap.min.css"
          integrity="sha384-rbsA2VBKQhggwzxH7pPCaAqO46MgnOM80zW1RWuH61DGLwZJEdK2Kadq2F9CUG65" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/default.min.css">
    <link href='http://fonts.googleapis.com/css?family=Roboto' rel='stylesheet' type='text/css'>
    <link href='http://fonts.googleapis.com/css?family=Roboto+Mono' rel='stylesheet' type='text/css'>
    <style>
        :root {
            --bs-font-sans-serif: Roboto, sans-serif;
        }
        pre, code {
            font-family: 'Roboto Mono', monospace;
        }
        pre {
            white-space: pre-wrap;
            white-space: -moz-pre-wrap;
            white-space: -pre-wrap;
            white-space: -o-pre-wrap;
            word-wrap: break-word;
            font-size: small;
        }
        @page {
            margin: 2rem;
            size: 21cm 29.7cm;
        }
        h1, h2, h3, h4, h5 {
            margin-top: 2rem;
            margin-bottom: 1.5rem;
        }
        hr {
            page-break-after: always;
            visibility: hidden;
        }
        html body * {
            page-break-inside: avoid;
        }
        html {
            font-family: Roboto, sans-serif;
        }
    </style>
</head>
<body>
<div class="container">
    <div class="row">
        <div class="col">
            <h1 id="emptyreport">Empty Report</h1>
<h1 id="summary">Summary</h1>
<p>This security report was conducted on 01/04/2020 at 01:30:10 (UTC+0).
A total of 0 issue(s) were found, 0 of which may require immediate attention.</p>
<p>This report is produced by running automated security scanning tools, which will likely not detect
all vulnerabilities present. <strong>It is not a replacement for a manual analysis of the application</strong>.</p>
<p>The following technical impacts may arise if an adversary successfully exploits one of the issues found by this scan.</p>
<h1 id="statistics">Statistics</h1>
<p>This report found issues with the following severities.</p>
<p><strong>Critical</strong>: 0 | <strong>High</strong> 0 | <strong>Medium</strong> 0 | <strong>Low</strong> 0 | <strong>Informational</strong> 0 | <strong>Unknown</strong> 0</p>
<p>To gain a better understanding of the severity levels please see <a href="#what-are-severity-levels">the appendix</a>.</p>
<h2 id="contents">Contents</h2>
<ul>
<li><a href="#summary">Summary</a><ul>
<li><a href="#statistics">Statistics</a></li></ul></li>
<li><a href="#overviewofissues">Overview of Issues</a></li>
<li><a href="#vulnerabilities">Vulnerabilities</a></li>
<li><a href="#additionalinformation">Additional Information</a><ul>
<li><a href="#whatareseveritylevels">What are severity levels?</a></li></ul></li>
</ul>
<h1 id="overviewofissues">Overview of Issues</h1>
<h1 id="vulnerabilities">Vulnerabilities</h1>
<h1 id="additionalinformation">Additional Information</h1>
<h2 id="whatareseveritylevels">What are severity levels?</h2>
<p>Issue severity is scored using the <a href="https://www.first.org/cvss/">Common Vulnerability Scoring System</a> (CVSS) where
such data is available. Severity levels do not represent the risk associated with an issue as risk depends on your
specific context. Severity scoring does however give an indication of the ease of exploitation and potential scope of an
attacks effect on an application.</p>
<h3 id="critical">Critical</h3>
<p>Exploitation will likely lead to an attacker gaining administrative access to the application and infrastructure that
supports it. Exploiting critical vulnerabilities is usually trivial and will generally not require prior access to the
application. <strong>A development team should aim to resolve these issues immediately by mitigating or directly resolving the
issue</strong>.</p>
<h3 id="high">High</h3>
<p>Exploitation could lead to an attacker gaining elevated access to the application and the infrastructure that supports
it. It is likely that an attacker will not find exploitation trivial. Such exploitation could lead to significant data
loss or downtime.</p>
<h3 id="medium">Medium</h3>
<p>Exploitation could lead to an attacker gaining limited access to the application. Exploiting vulnerabilities may require
an attacker to manipulate users to gain access to their credentials. Such exploitation could lead to limited data loss
or downtime.</p>
<h3 id="low">Low</h3>
<p>Exploitation will likely have very little impact on the application, and it is unlikely that an attacker will gain any
meaningful access to the application. Exploiting an issue of this severity will potentially require physical access to
the infrastructure that supports the application.</p>
<h3 id="informational">Informational</h3>
<p>While not part of the CVSS scoring specification, several security analysis tools use this severity level to indicate
that an issue is a matter of best practice. It is extremely unlikely that issues with this severity will lead to an
attacker gaining access to any application components.</p>
<h3 id="unknown">Unknown</h3>
<p>This severity level is used when the analysis tool used to perform a scan of the application does not associate any kind
of severity level with the issues or vulnerabilities it finds. Issues with an unknown severity should be investigated by
application developers and project stakeholders to establish the ease of exploitation, scope of any potential impact and
the specific risks associated.</p>
        </div>
    </div>
</div>
<script src="https://cdn.jsdelivr.net/npm/bootstrap@5.2.3/dist/js/bootstrap.bundle.min.js"
        integrity="sha384-kenU1KFdBIe4zVF0s0G1M5b4hcpxyD9F7jL+jjXkk+Q2h455rYXK/7HAuoJl+0I4"
        crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
<script>hljs.highlightAll();</script>
</body>
</html>
"
`;

exports[`RenderHTML an full report returns the expected report 1`] = `
"<!doctype html>
<html lang="en-gb">
<head>
    <title></title>
    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/bootstrap@5.2.3/dist/css/bootstrap.min.css"
          integrity="sha384-rbsA2VBKQhggwzxH7pPCaAqO46MgnOM80zW1RWuH61DGLwZJEdK2Kadq2F9CUG65" crossorigin="anonymous">
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/default.min.css">
    <link href='http://fonts.googleapis.com/css?family=Roboto' rel='stylesheet' type='text/css'>
    <link href='http://fonts.googleapis.com/css?family=Roboto+Mono' rel='stylesheet' type='text/css'>
    <style>
        :root {
            --bs-font-sans-serif: Roboto, sans-serif;
        }
        pre, code {
            font-family: 'Roboto Mono', monospace;
        }
        pre {
            white-space: pre-wrap;
            white-space: -moz-pre-wrap;
            white-space: -pre-wrap;
            white-space: -o-pre-wrap;
            word-wrap: break-word;
            font-size: small;
        }
        @page {
            margin: 2rem;
            size: 21cm 29.7cm;
        }
        h1, h2, h3, h4, h5 {
            margin-top: 2rem;
            margin-bottom: 1.5rem;
        }
        hr {
            page-break-after: always;
            visibility: hidden;
        }
        html body * {
            page-break-inside: avoid;
        }
        html {
            font-family: Roboto, sans-serif;
        }
    </style>
</head>
<body>
<div class="container">
    <div class="row">
        <div class="col">
            <h1 id="securityreportforapplication">Security Report for application</h1>
<h1 id="summary">Summary</h1>
<p>This security report was conducted on 01/04/2020 at 01:30:10 (UTC+0).
A total of 1 issue(s) were found, 0 of which may require immediate attention.</p>
<p>This report is produced by running automated security scanning tools, which will likely not detect
all vulnerabilities present. <strong>It is not a replacement for a manual analysis of the application</strong>.</p>
<p>The following technical impacts may arise if an adversary successfully exploits one of the issues found by this scan.</p>
<ul>
<li><strong>Availability</strong>: DoS: Crash, Exit, or Restart</li>
<li><strong>Confidentiality</strong>: Read Application Data</li>
</ul>
<h1 id="statistics">Statistics</h1>
<p>This report found issues with the following severities.</p>
<p><strong>Critical</strong>: 0 | <strong>High</strong> 0 | <strong>Medium</strong> 0 | <strong>Low</strong> 0 | <strong>Informational</strong> 0 | <strong>Unknown</strong> 1</p>
<p>To gain a better understanding of the severity levels please see <a href="#what-are-severity-levels">the appendix</a>.</p>
<h2 id="contents">Contents</h2>
<ul>
<li><a href="#summary">Summary</a><ul>
<li><a href="#statistics">Statistics</a></li></ul></li>
<li><a href="#overviewofissues">Overview of Issues</a><ul>
<li><a href="#CWE-248">Uncaught Exception</a></li></ul></li>
<li><a href="#vulnerabilities">Vulnerabilities</a><ul>
<li><a href="#unknownseverity">Unknown (1)</a></li></ul></li>
<li><a href="#additionalinformation">Additional Information</a><ul>
<li><a href="#whatareseveritylevels">What are severity levels?</a></li></ul></li>
</ul>
<h1 id="overviewofissues">Overview of Issues</h1>
<p><a id="CWE-248"></a></p>
<h2 id="uncaughtexception">Uncaught Exception</h2>
<p>An exception is thrown from a function, but it is not caught.</p>
<p>When an exception is not caught</p>
<h3 id="consequences">Consequences</h3>
<p>Using a vulnerability of this type an attacker may be able to affect the system in the following ways.</p>
<ul>
<li><strong>Availability</strong>: DoS: Crash, Exit, or Restart</li>
<li><strong>Confidentiality</strong>: Read Application Data</li>
</ul>
<blockquote>
  <p>An uncaught exception</p>
</blockquote>
<p>For more information see <a href="https://cwe.mitre.org/data/definitions/248.html">CWE-248</a>.</p>
<h1 id="vulnerabilities">Vulnerabilities</h1>
<h2 id="unknownseverity">Unknown Severity</h2>
<h3 id="testissuetitle">test-issue-title</h3>
<p><strong>Severity</strong>: <a href="#unknown">Unknown</a> | <strong>Type</strong>: code smell | <strong>Fix</strong>: Unknown | <strong>Found By</strong>: <a href="https://www.npmjs.com/package/test-scanner">test-scanner</a></p>
<p>test-issue-description</p>
<h4 id="evidence">Evidence</h4>
<p>The following evidence of this vulnerability was found in the application.</p>
<p><code>application/src/Report</code> (starting on line: 2)</p>
<pre><code class="python language-python">import
</code></pre>
<p><code>application/src/Report</code> (starting on line: 2)</p>
<pre><code class="python language-python">import
</code></pre>
<p><code>application/src/Report</code> (starting on line: 2)</p>
<pre><code class="python language-python">import
</code></pre>
<p><code>application/src/Report</code> (starting on line: 2)</p>
<pre><code class="python language-python">import
</code></pre>
<p><code>application/src/Report</code> (starting on line: 2)</p>
<pre><code class="python language-python">import
</code></pre>
<h5 id="references">References</h5>
<p><a href="#CWE-248">CWE-248</a></p>
<h1 id="additionalinformation">Additional Information</h1>
<h2 id="whatareseveritylevels">What are severity levels?</h2>
<p>Issue severity is scored using the <a href="https://www.first.org/cvss/">Common Vulnerability Scoring System</a> (CVSS) where
such data is available. Severity levels do not represent the risk associated with an issue as risk depends on your
specific context. Severity scoring does however give an indication of the ease of exploitation and potential scope of an
attacks effect on an application.</p>
<h3 id="critical">Critical</h3>
<p>Exploitation will likely lead to an attacker gaining administrative access to the application and infrastructure that
supports it. Exploiting critical vulnerabilities is usually trivial and will generally not require prior access to the
application. <strong>A development team should aim to resolve these issues immediately by mitigating or directly resolving the
issue</strong>.</p>
<h3 id="high">High</h3>
<p>Exploitation could lead to an attacker gaining elevated access to the application and the infrastructure that supports
it. It is likely that an attacker will not find exploitation trivial. Such exploitation could lead to significant data
loss or downtime.</p>
<h3 id="medium">Medium</h3>
<p>Exploitation could lead to an attacker gaining limited access to the application. Exploiting vulnerabilities may require
an attacker to manipulate users to gain access to their credentials. Such exploitation could lead to limited data loss
or downtime.</p>
<h3 id="low">Low</h3>
<p>Exploitation will likely have very little impact on the application, and it is unlikely that an attacker will gain any
meaningful access to the application. Exploiting an issue of this severity will potentially require physical access to
the infrastructure that supports the application.</p>
<h3 id="informational">Informational</h3>
<p>While not part of the CVSS scoring specification, several security analysis tools use this severity level to indicate
that an issue is a matter of best practice. It is extremely unlikely that issues with this severity will lead to an
attacker gaining access to any application components.</p>
<h3 id="unknown">Unknown</h3>
<p>This severity level is used when the analysis tool used to perform a scan of the application does not associate any kind
of severity level with the issues or vulnerabilities it finds. Issues with an unknown severity should be investigated by
application developers and project stakeholders to establish the ease of exploitation, scope of any potential impact and
the specific risks associated.</p>
        </div>
    </div>
</div>
<script src="https://cdn.jsdelivr.net/npm/bootstrap@5.2.3/dist/js/bootstrap.bundle.min.js"
        integrity="sha384-kenU1KFdBIe4zVF0s0G1M5b4hcpxyD9F7jL+jjXkk+Q2h455rYXK/7HAuoJl+0I4"
        crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
<script>hljs.highlightAll();</script>
</body>
</html>
"
`;
